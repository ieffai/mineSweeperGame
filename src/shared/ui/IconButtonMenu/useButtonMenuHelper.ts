import { useEffect, useId, useRef, useState } from 'react';

export function useButtonMenuHelper(isOpen?: boolean) {
  const id = useId();
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);
  const btnRef = useRef<HTMLButtonElement>(null);

  useEffect(() => {
    if (isOpen) {
      setAnchorEl(btnRef.current);
    } else {
      setAnchorEl(null);
    }
  }, [isOpen]);

  return { id, open, btnRef, anchorEl };
}
