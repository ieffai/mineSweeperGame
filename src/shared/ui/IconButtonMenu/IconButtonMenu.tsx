import { Popover } from '@mui/material';
import type { PopoverProps } from '@mui/material';
import type { PopoverOrigin } from '@mui/material';
import type { TransitionProps } from '@mui/material/transitions';
import { useButtonMenuHelper } from './useButtonMenuHelper';
import type { ButtonProps } from '../Button';
import type { ReactNode } from 'react';
import { IconButton } from '../IconButton';

export interface IconButtonMenuProps {
  isOpen: boolean;
  onClick: ButtonProps['onClick'];
  onClose: PopoverProps['onClose'];
  icon?: ReactNode;
  buttonProps?: Omit<ButtonProps, 'onClick, children'>;
  popoverProps?: {
    anchorOrigin?: PopoverOrigin;
    transformOrigin?: PopoverOrigin;
    TransitionProps?: TransitionProps;
  };
  children: ReactNode;
}

export const IconButtonMenu = ({
  isOpen,
  onClick,
  onClose,
  icon,
  buttonProps,
  popoverProps,
  children,
}: IconButtonMenuProps) => {
  const { anchorEl, btnRef, id, open } = useButtonMenuHelper(isOpen);

  return (
    <>
      <IconButton
        ref={btnRef}
        size="medium"
        {...buttonProps}
        onClick={onClick}
        aria-haspopup="true"
        aria-controls={id}
        aria-expanded={open ? 'true' : undefined}
        sx={{
          ...buttonProps?.sx,
        }}
      >
        {icon}
      </IconButton>
      <Popover
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'left',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'left',
        }}
        {...popoverProps}
        id={id}
        open={open}
        anchorEl={anchorEl}
        onClose={onClose}
      >
        {children}
      </Popover>
    </>
  );
};
