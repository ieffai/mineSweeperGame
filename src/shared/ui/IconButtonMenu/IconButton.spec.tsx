import { render } from '@testing-library/react';
import { vi } from 'vitest';
import { IconButtonMenu } from './IconButtonMenu';

describe('IconButtonMenu', () => {
  it('should render successfully', () => {
    const mockOnClick = vi.fn();
    const mockOnClose = vi.fn();

    const { asFragment } = render(
      <IconButtonMenu isOpen={false} onClick={mockOnClick} onClose={mockOnClose} icon={<span>ðŸ”§</span>}>
        <div>Menu content</div>
      </IconButtonMenu>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render when open', () => {
    const mockOnClick = vi.fn();
    const mockOnClose = vi.fn();

    const { asFragment } = render(
      <IconButtonMenu isOpen={true} onClick={mockOnClick} onClose={mockOnClose} icon={<span>ðŸ”§</span>}>
        <div>Menu content</div>
      </IconButtonMenu>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
