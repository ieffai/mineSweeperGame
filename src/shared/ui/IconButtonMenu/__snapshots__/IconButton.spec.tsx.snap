// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IconButtonMenu > should render successfully 1`] = `
<DocumentFragment>
  <button
    aria-controls="Â«r0Â»"
    aria-haspopup="true"
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
    tabindex="0"
    type="button"
  >
    <span>
      ğŸ”§
    </span>
  </button>
</DocumentFragment>
`;

exports[`IconButtonMenu > should render when open 1`] = `
<DocumentFragment>
  <button
    aria-controls="Â«r2Â»"
    aria-expanded="true"
    aria-haspopup="true"
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
    tabindex="0"
    type="button"
  >
    <span>
      ğŸ”§
    </span>
  </button>
</DocumentFragment>
`;
